### PR TITLE
Add table block column alignment

### LINKED ISSUE #15823: Table block: Text alignment (none, left, center, right)
The Table block currently does not allow for text alignment within a cell. Users can benefit from text alignment within a cell when using numbers, dates, money. Often times these values are aligned right. This can't be done currently.

These options:

Screen Shot 2019-05-24 at 1 34 59 PM

Should exist in the Table block toolbar.

Screen Shot 2019-05-24 at 1 37 37 PM

This was initially noted in https://github.com/WordPress/gutenberg/issues/6923 by @helen. I'm pulling it into it's own issue.

### PR DESCRIPTION
**Pull request by talldan**
## Description
Adds column alignment options to the table block. Closes #15823

**Dev decisions**
Code-wise, the part where I've had to break from the norm is by using a `data-align` attribute to store the alignment value for a cell. This is because the cell attributes are deeply nested and sourced using a `query` matcher, so normal attributes can't be used.

Applying an alignment adds a class name and `data-align` attribute to each table cell in a column. There aren't many alternatives to this approach that I can think of, apart from using inline styles. `colgroup`/`col` elements don't support text align, so they're out of the question.

## How has this been tested?
Added unit and e2e tests.

Manual testing steps:
1. Add a table block to a post
2. Add some text to the first column of the table.
3. Ensure one of the cells in the first column is selected
4. From the toolbar try each of the column alignment options
5. Verify that text in the each cell in the column is aligned in the editor
6. Preview the post
7. Verify that the text is aligned when viewing the post.

## Screenshots 
**Editor**
Screen Shot 2019-07-30 at 10 46 50 am

**Post**
Screen Shot 2019-07-30 at 10 47 45 am

## Types of changes

New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by mapk** [comment]
I just tested this and it works wonderfully! Thanks, @talldan! 👍 from a design perspective. :shipit: 

Screen Shot 2019-06-15 at 7 18 06 AM

**Comment by talldan** [comment]
Thanks! I think I'll switch this to using classnames as #16035 does, as that's now been merged.

**Comment by mtias** [comment]
Should this apply to the whole table, per column, or on individual cells?

**Comment by mapk** [comment]
> Should this apply to the whole table, per column, or on individual cells?

Great question, @mtias. Right now the block only allows selection of a single cell at a time. I would love to see selections happen across rows and columns. Once this works, we can look at alignment across these.

**Comment by talldan** [comment]
I've refactored this to use classes, so it's ready for a code review 👍

File: packages/block-library/src/table/block.json [hunk 1]
```diff
@@ -29,6 +29,11 @@
 							"type": "string",
 							"default": "td",
 							"source": "tag"
+						},
+						"align": {
+							"type": "string",
+							"source": "attribute",
+							"attribute": "data-align"
```
Comment (youknowriad) [inline packages/block-library/src/table/block.json]: Interesting. since it applied as a class, makes me wonder if at some point, we'd want "classNames" sources. That said, it can be complex and I know it's a deliberate decision to restrict the number of sources available. css @aduth

File: packages/block-library/src/table/edit.js [hunk 2]
```diff
@@ -465,6 +502,11 @@ export class TableEdit extends Component {
 							controls={ this.getTableControls() }
 						/>
 					</Toolbar>
+					<AlignmentToolbar
+						value={ this.getCellAlignment() }
+						onChange={ ( nextAlign ) => this.onChangeCellAlignment( nextAlign ) }
+						isCollapsed={ true }
```
Comment (youknowriad) [inline packages/block-library/src/table/edit.js]: Nit: `={ true }` can be dropped.

File: packages/block-library/src/table/style.scss [hunk 3]
```diff
@@ -76,6 +76,18 @@
 		th,
 		td {
 			border-color: transparent;
+
+			&.has-text-align-left {
```
Comment (youknowriad) [inline packages/block-library/src/table/style.scss]: Do we really need these classNames to be defined here? I thought these were meant to be global classNames?

**Review by youknowriad — COMMENTED** [review]
This seesm to be in a good shape. I left some minor comments.

**Review by ellatrix — CHANGES_REQUESTED** [review]
I have some strong opinion about the user experience of this, sorry.

1. We should only allow table cells to be left- or right-aligned. The button can be a simple toggle. Center alignment in tables doesn't make so much sense. What kind of data will you align like that? The general rule of thumb is that textual data should be left-aligned, and numerical data should be right-aligned.
2. Another table "rule" is that each column should have the same alignment. This not only looks better,  it easier for the user to align the entire column at once.

**Comment by mtias** [comment]
I think center alignment can be perfectly fine in some cases, and we should not dictate that. However, I think that single cell alignment is a bit excessive. I'd rather have a text alignment toolbar button that applies to the whole table at once (since columns is a bit more complicated).

**Comment by ellatrix** [comment]
We can always add center alignment if it's much requested. It's easy to add, but hard to take away. Center alignment only makes tables much harder to scan.

Why is applying it to the column complicated? Aligning an entire table seems really strange to me. A table usually doesn't contain only numerical data (at the very least, the row headings won't be numerical).

File: packages/block-library/src/table/block.json (hunk 1 above)
Comment (aduth) [reply packages/block-library/src/table/block.json]: Yeah, I think we'd considered once before a source serialized as the presence of a class name. Personally I don't have much of an issue with the `data-` attributes; they can be styled just as well, are a bit easier to parse, and fit well as serving the purpose of a data attribute.

>Custom data attributes are intended to store custom data, state, annotations, and similar, private to the page or application, for which there are no more appropriate attributes or elements.

https://html.spec.whatwg.org/multipage/dom.html#embedding-custom-non-visible-data-with-the-data-*-attributes

That being said, I'm not really sure why we bother to serialize this in the HTML at all, if we're not using it for anything like styling? Why not just embed it within the comment delimiters (i.e. omit `source`)?

**Comment by talldan** [comment]
I'm not sure I see much difference between other formatting options and alignment. For example, I'm unlikely to want to apply `bold` formatting to just a single cell, instead I'd be more likely to select the entire header or footer row and choose bold formatting. However, this is the way it works in the table block today.

I started looking at column/row/table selection in another PR (https://github.com/WordPress/gutenberg/pull/16493). It's in a very early state, needs some design and accessibility work. Also needs some more work to make it actually work. I think it's something that'll be required as we start to add more options to the table block that can be applied to an individual or multiple cells. It would also be a good first stepping stone to more freeform multiple selection of cells.

I think without that, clicking align and having it apply to an entire column or table when there's only a single cell selected feels like an unexpected outcome. It'd be better to put the effort into #16493 in my opinion.

File: packages/block-library/src/table/edit.js (hunk 2 above)
Comment (talldan) [reply packages/block-library/src/table/edit.js]: 👍

File: packages/block-library/src/table/style.scss (hunk 3 above)
Comment (talldan) [reply packages/block-library/src/table/style.scss]: Well spotted, I didn't realise there were global styles. Removed those styles.

File: packages/block-library/src/table/block.json (hunk 1 above)
Comment (talldan) [reply packages/block-library/src/table/block.json]: @aduth As this attribute is in a `query`, my understanding is that `source` can't be omitted. A `classNames` source could be useful in this use case.

An argument against `data-` attributes is that it's double storing data, but then it's not too much of a problem with the reactive nature of blocks, it's not like there's extra business logic to keep them in sync.

**Comment by talldan** [comment]
Have rebased and updated this based the code review feedback (thanks for the review!).

I've added the 'Needs Decision' label, as a decision will have to be made on the how alignment is applied (per cell, column or table). The debate on that is in the four posts above this one for anyone reading.

**Comment by talldan** [comment]
I've made a branch [`try/table-multi-selection-with-alignments`](https://github.com/WordPress/gutenberg/compare/try/table-multi-selection-with-alignments?expand=1) that merges this PR with #16493. Here's a demo:

table-alignment

Pretty easy to modify any cell attributes across selections with this code. The tough bit is going to be applying rich text formatting across cells. I'm not entirely sure how to achieve that. Any ideas @ellatrix?

**Comment by mapk** [comment]
> We should only allow table cells to be left- or right-aligned. The button can be a simple toggle. Center alignment in tables doesn't make so much sense. What kind of data will you align like that? 

@ellatrix, there are a couple examples that come to mind. Some of which can be seen in this graphic from wordpress.com/pricing below:

Screen Shot 2019-07-24 at 9 44 43 AM

1. I look forward to a time when we can add images/icons to the Table block. In which case, center-aligning checkboxes like this makes sense.
2. Center aligning the column headers also makes sense depending on the design.

> Another table "rule" is that each column should have the same alignment. This not only looks better, it easier for the user to align the entire column at once.

I love this and agree that moving these alignments to column-wide settings would be great!

@talldan while I like the per-cell adjustements I can make, I do long for the column-based alignments and will test that other PR today. One thing I noted is in @jasmussen's PR https://github.com/WordPress/gutenberg/pull/16557#issuecomment-514639436 he uses an arrow icon to indicate the dropdown of block alignments. Should we do the same here for the text alignments?

**Comment by talldan** [comment]
@mapk I think we'll need to solve that a cell is shown selected when aligning, because that'll be confusing. I can look into highlighting the column that will be aligned when the alignment icon button is hovered, and also change the label text to 'Align column ...'. 

I think that would make it clear enough.

**Comment by talldan** [comment]
I've updated this PR so that it now aligns entire columns. 

Tests are all green, so it should be ready for review.

File: packages/block-library/src/table/editor.scss [hunk 4]
```diff
@@ -41,9 +41,19 @@
 
 	td.is-selected,
 	th.is-selected {
-		border-color: $blue-medium-500;
-		box-shadow: inset 0 0 0 1px $blue-medium-500;
-		border-style: double;
+		position: relative;
```
Comment (ellatrix) [inline packages/block-library/src/table/editor.scss]: What is the reason behind these changes? What does it have to do with alignment?

File: packages/block-library/src/table/state.js [hunk 5]
```diff
@@ -26,105 +28,186 @@ export function createTable( {
...
+
+/**
+ * Returns whether the cell at `cellLocation` is included in the selection `selection`.
+ *
+ * @param {Object} cellLocation An object containing cell location properties.
+ * @param {Object} selection    An object containing selection properties.
+ *
+ * @return {boolean} True if the cell is selected, false otherwise.
+ */
+export function isCellSelected( cellLocation, selection ) {
```
Comment (ellatrix) [inline packages/block-library/src/table/state.js]: Why not have `isCellSelected` and `isColumnSelected`?

File: packages/block-library/src/table/state.js [hunk 6]
```diff
@@ -26,105 +28,186 @@ export function createTable( {
...
 }
 
 /**
  * Inserts a row in the table state.
  *
- * @param {Object} state            Current table state.
- * @param {string} options.section  Section in which to insert the row.
- * @param {number} options.rowIndex Row index at which to insert the row.
+ * @param {Object} state                Current table state.
+ * @param {string} options.sectionName  Section in which to insert the row.
```
Comment (ellatrix) [inline packages/block-library/src/table/state.js]: Looks like there's a space too much?

File: packages/block-library/src/table/state.js [hunk 7]
```diff
@@ -26,105 +28,186 @@ export function createTable( {
...
  */
 export function insertRow( state, {
-	section,
+	sectionName,
 	rowIndex,
 	columnCount,
 } ) {
-	const cellCount = columnCount || state[ section ][ 0 ].cells.length;
+	const firstRow = getFirstRow( state );
+	const cellCount = columnCount || get( firstRow, [ 'cells', 'length' ], 2 );
```
Comment (ellatrix) [inline packages/block-library/src/table/state.js]: Why does it default to 2?

File: packages/block-library/src/table/state.js [hunk 8]
```diff
@@ -26,105 +28,186 @@ export function createTable( {
...
+		[ sectionName ]: [
+			...state[ sectionName ].slice( 0, rowIndex ),
 			{
-				cells: times( cellCount, () => ( {
-					content: '',
-					tag: section === 'head' ? 'th' : 'td',
-				} ) ),
+				cells: times( cellCount, ( index ) => {
+					const firstCellInColumn = get( firstRow, [ 'cells', index ], {} );
+					const inheritedAttributes = pick( firstCellInColumn, INHERITED_COLUMN_ATTRIBUTES );
```
Comment (ellatrix) [inline packages/block-library/src/table/state.js]: Why is alignment inherited?

File: packages/block-library/src/table/block.json (hunk 1 above)
Comment (ellatrix) [reply packages/block-library/src/table/block.json]: Why use the class at all? Can we not just use the data attribute?

File: packages/block-library/src/table/state.js [hunk 9]
```diff
@@ -26,105 +28,186 @@ export function createTable( {
...
+ * @param {number} options.attributeName The name of the attribute to get the value of.
+ *
+ * @return {*} The attribute value.
+ */
+export function getCellAttribute( state, {
+	sectionName,
 	rowIndex,
 	columnIndex,
-	content,
+	attributeName,
```
Comment (ellatrix) [inline packages/block-library/src/table/state.js]: I think this would read better if `attributeName` is a separate argument. E.g. `getCellAttribute( state, 'align', cell )`.

File: packages/block-library/src/table/edit.js [hunk 10]
```diff
@@ -156,14 +179,73 @@ export class TableEdit extends Component {
...
-		setAttributes( updateCellContent( attributes, {
-			section,
-			rowIndex,
-			columnIndex,
-			content,
-		} ) );
+		setAttributes( updateSelectedCell(
+			attributes,
+			selectedCell,
+			( cellAttributes ) => ( { ...cellAttributes, content } ),
```
Comment (ellatrix) [inline packages/block-library/src/table/edit.js]: Why do we need a callback here? Can this not be done directly by `updateSelectedCell`?

File: packages/block-library/src/table/state.js (hunk 5 above)
Comment (talldan) [reply packages/block-library/src/table/state.js]: `isColumnSelected` would only be used in one place, so it seems unnecessary at this stage to have two separate functions.

I'm exploring this more in #16493, where I do have separate functions for different types of selections as well as an `isCellSelected` for easily checking whether a given cell is in any type of selection:
https://github.com/WordPress/gutenberg/blob/ec1aa436ec4e4ce641799ec96e9bc04862405da9/packages/block-library/src/table/state.js#L221-L244

It'll be an easy refactor from this to that.

File: packages/block-library/src/table/state.js (hunk 8 above)
Comment (talldan) [reply packages/block-library/src/table/state.js]: When adding a new row, the newly added cells will keep the same alignment as the rest of their adjacent column.

As already mentioned in this PR, it seems like all cells in a column should have the same alignment, so this maintains that experience for the user. In my testing it felt like the right way to go.

File: packages/block-library/src/table/state.js (hunk 7 above)
Comment (talldan) [reply packages/block-library/src/table/state.js]: That's a fair comment. I thought `2` because that's what the table block defaults to when created.

Maybe this should instead be considered unexpected and the `insertRow` function should return early with no effect if `cellCount` is undefined.

File: packages/block-library/src/table/edit.js (hunk 10 above)
Comment (talldan) [reply packages/block-library/src/table/edit.js]: It might be overkill, but I feel it's a more useful API. Being able to read the cell's attributes provides an opportunity for doing things like toggling values or incrementing/decrementing values when updating them.

File: packages/block-library/src/table/editor.scss (hunk 4 above)
Comment (talldan) [reply packages/block-library/src/table/editor.scss]: It is a better way to apply a cell border than box-shadow, but not needed for this PR so I've removed these changes.

File: packages/block-library/src/table/block.json (hunk 1 above)
Comment (talldan) [reply packages/block-library/src/table/block.json]: There are some global styles for text alignment:
https://github.com/WordPress/gutenberg/blob/4ac7dc9d2d93a1b3ab81b44a6fbd89952a44915e/packages/block-library/src/style.scss#L150-L161

Seems better to be able to use those and not have extra styles in the table block.

**Comment by paaljoachim** [comment]
I assume one can then multi select cells in a column and have all of them align in one go.

File: packages/block-library/src/table/state.js (hunk 7 above)
Comment (talldan) [reply packages/block-library/src/table/state.js]: I changed it to return early with unchanged state if `cellCount` is unknown.

File: packages/block-library/src/table/state.js (hunk 9 above)
Comment (talldan) [reply packages/block-library/src/table/state.js]: Agreed it could be improved. I've changed it to `getCellAttribute( state, cell, 'align' )`.

**Comment by talldan** [comment]
> I assume one can then multi select cells in a column and have all of them align in one go.

@paaljoachim Sort of, the table still only has column-based selection, but alignments apply to the column that the selected cell is in.

File: packages/block-library/src/table/state.js (hunk 5 above)
Comment (ellatrix) [reply packages/block-library/src/table/state.js]: That's fine, but then I wouldn't call it `isCellSelected`? That's misleading if you're checking if a column is selected. `isCellSelected( ..., 'column' )` doesn't make sense?

File: packages/block-library/src/table/state.js (hunk 7 above)
Comment (ellatrix) [reply packages/block-library/src/table/state.js]: I think so too. If there's no amount of rows, there's nothing to be done? If you'd like to keep a default, I feel like it shouldn't be hardcoded here, but set by the component instead (which knows the default row count).

File: packages/block-library/src/table/state.js (hunk 8 above)
Comment (ellatrix) [reply packages/block-library/src/table/state.js]: Sounds good

File: packages/block-library/src/table/state.js (hunk 9 above)
Comment (ellatrix) [reply packages/block-library/src/table/state.js]: Great! It seems like a good idea to keep `cellLocation` as a separate entity. Could this object be documented somewhere to the type can be used in the docs of all these functions?

File: packages/e2e-tests/specs/blocks/table.test.js [hunk 11]
```diff
@@ -10,6 +15,17 @@ import {
...
 const createButtonSelector = "//div[@data-type='core/table']//button[text()='Create Table']";
 
+/**
+ * Utility function for changing the selected cell alignment.
+ *
+ * @param {string} align The alignment (one of 'left', 'center', or 'right').
+ */
+async function changeCellAlignment( align ) {
+	await clickBlockToolbarButton( 'Change column alignment' );
+	const alignButton = await page.$x( `//button[text()='Align Column ${ capitalize( align ) }']` );
```
Comment (ellatrix) [inline packages/e2e-tests/specs/blocks/table.test.js]: Can the `clickButton` utility function be used here?

File: packages/e2e-tests/specs/blocks/table.test.js [hunk 12]
```diff
@@ -154,4 +170,40 @@ describe( 'Table', () => {
...
+	it( 'allows columns to be aligned', async () => {
+		await insertBlock( 'Table' );
+
+		const [ columnCountLabel ] = await page.$x( "//div[@data-type='core/table']//label[text()='Column Count']" );
+		await columnCountLabel.click();
+		await page.keyboard.press( 'Backspace' );
+		await page.keyboard.type( '4' );
+
+		// Create the table.
+		const [ createButton ] = await page.$x( createButtonSelector );
```
Comment (ellatrix) [inline packages/e2e-tests/specs/blocks/table.test.js]: Also here, could `clickButton` be used?

**Review by ellatrix — APPROVED** [review]

File: packages/block-library/src/table/state.js (hunk 5 above)
Comment (talldan) [reply packages/block-library/src/table/state.js]: Hmm, yeah, that's good feedback. I can work on a few naming improvements in future PRs as this functionality develops. Something like `isCellWithinSelection` might be more appropriate.

Thanks for the reviews btw!

File: packages/e2e-tests/specs/blocks/table.test.js (hunk 11 above)
Comment (talldan) [reply packages/e2e-tests/specs/blocks/table.test.js]: Genuinely didn't know that existed. Will do a follow up PR to tidy up the tests.
